File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/superclasses/RobotModule.java
package org.firstinspires.ftc.teamcode.superclasses;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;

@SuppressWarnings("EmptyMethod")
public abstract class RobotModule {

    protected LinearOpMode opMode = null;

    public void setOpMode(LinearOpMode opMode) {
        this.opMode = opMode;
    }

    public void initialize() {
    }

    public final void initialize(LinearOpMode opMode) {
        setOpMode(opMode);
        initialize();
    }

    public void start() {
    }

    public void updateAll() {
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/superclasses/RobotModule.kt
package org.firstinspires.ftc.teamcode.superclasses

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode

abstract class RobotModule {
    lateinit var opMode: LinearOpMode

    open fun initialize() {}

    fun initialize(opMode: LinearOpMode) {
        this.opMode = opMode
        initialize()
    }

    open fun start() {}
    open fun updateAll() {}

    open val status: String
    get() = this.toString()
}